### ISSUE TITLE
Python 3.7 test failures

### ISSUE THREAD

**Issue by hroncok** [eventlet/eventlet#502]
After https://github.com/eventlet/eventlet/issues/475 is fixed on greenlet side, there are still test failures on python 3.7:

```
======================================================================
ERROR: test_connect_ssl (tests.api_test.TestApi)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/api_test.py", line 67, in test_connect_ssl
    server_side=True
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/convenience.py", line 158, in wrap_ssl
    return wrap_ssl_impl(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_wrap_ssl (tests.convenience_test.TestServe)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/convenience_test.py", line 121, in test_wrap_ssl
    server_side=True)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/convenience.py", line 158, in wrap_ssl
    return wrap_ssl_impl(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_wrap_socket (tests.greenio_test.TestGreenSocket)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/greenio_test.py", line 476, in test_wrap_socket
    ssl.wrap_socket(sock)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_duplex_response (tests.ssl_test.SSLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 45, in test_duplex_response
    sock = listen_ssl_socket()
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 22, in listen_ssl_socket
    **kwargs
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_greensslobject (tests.ssl_test.SSLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 192, in test_greensslobject
    listener = listen_ssl_socket()
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 22, in listen_ssl_socket
    **kwargs
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_no_handshake_block_accept_loop (tests.ssl_test.SSLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 210, in test_no_handshake_block_accept_loop
    listener = listen_ssl_socket()
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 22, in listen_ssl_socket
    **kwargs
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_receiving_doesnt_block_if_there_is_already_decrypted_buffered_data (tests.ssl_test.SSLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 278, in test_receiving_doesnt_block_if_there_is_already_decrypted_buffered_data
    listener = listen_ssl_socket()
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 22, in listen_ssl_socket
    **kwargs
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().
-------------------- >> begin captured stdout << ---------------------
Trying <function SSLTest.test_receiving_doesnt_block_if_there_is_already_decrypted_buffered_data.<locals>.recv at 0x7f72781dea60>...

--------------------- >> end captured stdout << ----------------------

======================================================================
ERROR: test_recv_after_ssl_connect (tests.ssl_test.SSLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 93, in test_recv_after_ssl_connect
    sock = listen_ssl_socket()
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 22, in listen_ssl_socket
    **kwargs
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_regression_gh_17 (tests.ssl_test.SSLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 202, in test_regression_gh_17
    sock = ssl.wrap_socket(socket.socket())
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: SSL socket.sendall() busy loop
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 173, in test_sendall_cpu_usage
    server_sock = listen_ssl_socket()
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 22, in listen_ssl_socket
    **kwargs
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_ssl_close (tests.ssl_test.SSLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 63, in test_ssl_close
    sock = listen_ssl_socket()
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 22, in listen_ssl_socket
    **kwargs
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_ssl_connect (tests.ssl_test.SSLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 78, in test_ssl_connect
    sock = listen_ssl_socket()
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 22, in listen_ssl_socket
    **kwargs
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_ssl_unwrap (tests.ssl_test.SSLTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/ssl_test.py", line 140, in test_ssl_unwrap
    client_ssl = ssl.wrap_socket(client)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_ssl_sending_messages (tests.websocket_test.TestWebSocketSSL)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/websocket_test.py", line 541, in test_ssl_sending_messages
    server_side=True)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/convenience.py", line 158, in wrap_ssl
    return wrap_ssl_impl(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_012_ssl_server (tests.wsgi_test.TestHttpd)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/wsgi_test.py", line 451, in test_012_ssl_server
    server_side=True)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/convenience.py", line 158, in wrap_ssl
    return wrap_ssl_impl(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_013_empty_return (tests.wsgi_test.TestHttpd)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/wsgi_test.py", line 473, in test_013_empty_return
    server_side=True)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/convenience.py", line 158, in wrap_ssl
    return wrap_ssl_impl(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_017_ssl_zeroreturnerror (tests.wsgi_test.TestHttpd)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/wsgi_test.py", line 578, in test_017_ssl_zeroreturnerror
    server_side=True)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/convenience.py", line 158, in wrap_ssl
    return wrap_ssl_impl(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
ERROR: test_028_ssl_handshake_errors (tests.wsgi_test.TestHttpd)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/wsgi_test.py", line 1110, in test_028_ssl_handshake_errors
    server_side=True)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/convenience.py", line 158, in wrap_ssl
    return wrap_ssl_impl(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 379, in wrap_socket
    return GreenSSLSocket(sock, *a, **kw)
  File "/builddir/build/BUILD/eventlet-0.23.0/eventlet/green/ssl.py", line 68, in __init__
    ca_certs, do_handshake_on_connect and six.PY2, *args, **kw)
  File "/usr/lib64/python3.7/ssl.py", line 788, in __init__
    f"{self.__class__.__name__} does not have a public "
TypeError: GreenSSLSocket does not have a public constructor. Instances are returned by SSLContext.wrap_socket().

======================================================================
FAIL: test_line (tests.debug_test.TestSpew)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/debug_test.py", line 39, in test_line
    assert "f=<frame object at" in output
AssertionError

======================================================================
FAIL: test_line_global (tests.debug_test.TestSpew)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/builddir/build/BUILD/eventlet-0.23.0/tests/debug_test.py", line 62, in test_line_global
    assert "f=<frame object at" in output
AssertionError

----------------------------------------------------------------------
Ran 671 tests in 49.917s

FAILED (SKIP=95, errors=18, failures=2)

```

GreenSSLSocket is in:

https://github.com/eventlet/eventlet/blob/acd8b65a06e90dbeb1ea416d1cd5922a0727ebd2/eventlet/green/ssl.py#L368-L369

**Comment by hroncok** [comment]
I guess the entire `__init__` of `GreenSSLSocket` needs to get replaced with `_create`. I've been trying but failed.

**Comment by skaparis** [comment]
It's seems that from python 3.7 its not a good idia to sublcass ssl.socket. I also facing this problem. Its not clear yet how to make workaround.

**Comment by madanh** [comment]
I'm affected by this when trying to use flask-socketio

**Comment by yogevyuval** [comment]
Fixed in eventlet 0.26.0
